### PR TITLE
GODRIVER-2265 Make more correct assertions in SDAM heartbeat test.

### PR DESCRIPTION
**Pull request by matthewdale**
[GODRIVER-2265](https://jira.mongodb.org/browse/GODRIVER-2265)

The test `TestSDAMProse/heartbeats_processed_more_frequently` fails intermittently on the Evergreen waterfall. Some assertions made in that test are only accidentally working:
* The previous assertion that server version <4.4 processes a different number of operations for all heartbeat intervals is not correct. The discrepancy between the number of operations processed between server versions <4.4 and >=4.4 comes from the "minimum heartbeat interval" of 500ms that only applies to non-streaming heartbeats.
* The previous assertion that a `Client` should send a heartbeat exactly every interval is not correct. A `Client` waits _at least_ the specified interval between heartbeats, and does not guarantee exactly when the next heartbeat will run (the test asserts that heartbeats can occur at up to 2 times the specified interval).

Correct the assertions made in the `TestSDAMProse/heartbeats_processed_more_frequently` test and hopefully improve the reliability of the test in some circumstances.

### CODE REVIEW

File: mongo/integration/sdam_prose_test.go [hunk 1]
```diff
@@ -22,34 +22,51 @@ func TestSDAMProse(t *testing.T) {
...
+		duration := 2 * time.Second
 
 		numNodes := len(options.Client().ApplyURI(mtest.ClusterURI()).Hosts)
-		timeDuration := 250 * time.Millisecond
-		numExpectedResponses := numNodes * int(timeDuration/lowHeartbeatFrequency)
-		if mtest.CompareServerVersions(mtest.ServerVersion(), "4.4") >= 0 {
-			numExpectedResponses = numNodes * (2*int(timeDuration/lowHeartbeatFrequency) + 1)
-		}
+		maxExpected := numNodes * (1 + 2*int(duration/heartbeatInterval))
+		minExpected := numNodes * (1 + 2*int(duration/(heartbeatInterval*2)))
```
Comment (benjirewis) [inline mongo/integration/sdam_prose_test.go:55]: `maxExpected` and `minExpected` should always be `numNodes * 9` and `numNodes * 5` respectively, right? Leaving this math in seems clearer to represent the logic, but just wanted to clarify that we're expecting between 5 and 9 heartbeats/handshakes for each node within 2 seconds when `heartbeatInterval` is set to 500ms.

**Review by benjirewis — COMMENTED** [review]
Thanks for updating this test! If your assertions about the behavior of `heartbeatInterval` are correct, this looks good. Is `heartbeatInterval` described as a rate-limiting option (the `Client` will wait _at least_ 500ms between heartbeats) somewhere in the spec?

**Comment by matthewdale** [comment]
The statement

> the Client will wait _at least_ 500ms between heartbeats

is based on the behavior of timers in Go, not based on the spec. For example, the description of [time.NewTimer](https://pkg.go.dev/time#NewTimer) says it

> will send the current time on its channel after at least duration d

Basically, Go timers don't guarantee exactly when something will happen, only that it won't happen before a specific time. The previous test attempted to assert behavior closer to exact timing, but failed when the Go runtime couldn't schedule those heartbeats exactly on time (maybe because the Evergreen host CPU was busy).

There is a rate-limit described in the spec, [minHeartbeatFrequencyMS](https://github.com/mongodb/specifications/blob/1823a4e95dc9e6509734c959b2e933b9a773f9e5/source/server-discovery-and-monitoring/server-monitoring.rst#minheartbeatfrequencyms).

> If a client frequently rechecks a server, it MUST wait at least minHeartbeatFrequencyMS milliseconds since the previous check ended

That describes a similar behavior to how timers in Go work, so they're actually both relevant, but my original intent was to express the limitation of Go timers.

File: mongo/integration/sdam_prose_test.go (hunk 1 above)
Comment (matthewdale) [reply mongo/integration/sdam_prose_test.go:55]: That's correct.

**Review by benjirewis — APPROVED** [review]
Gotcha, thanks for the explanation. Looks good to me.

**Pull request merged**